src/ripple/rpc/impl/NFTokenID.cpp

```diff
@@ -0,0 +1,192 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/rpc/NFTokenID.h>
+
+#include <ripple/app/ledger/LedgerMaster.h>
+#include <ripple/app/ledger/OpenLedger.h>
+#include <ripple/app/misc/Transaction.h>
+#include <ripple/ledger/View.h>
+#include <ripple/net/RPCErr.h>
+#include <ripple/protocol/AccountID.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/rpc/Context.h>
+#include <ripple/rpc/impl/RPCHelpers.h>
+#include <boost/algorithm/string/case_conv.hpp>
+
+namespace ripple {
+namespace RPC {
+
+bool
+canHaveNFTokenID(
+    std::shared_ptr<STTx const> const& serializedTx,
+    TxMeta const& transactionMeta)
+{
+    if (!serializedTx)
+        return false;
+
+    TxType const tt{serializedTx->getTxnType()};
+    if (tt != ttNFTOKEN_MINT && tt != ttNFTOKEN_ACCEPT_OFFER &&
+        tt != ttNFTOKEN_CANCEL_OFFER)
+        return false;
+
+    // if the transaction failed nothing could have been delivered.
+    if (transactionMeta.getResultTER() != tesSUCCESS)
+        return false;
+
+    return true;
+}
+
+void
+getNFTokenIDFromPage(
+    TxMeta const& transactionMeta,
+    std::vector<uint256>& tokenIDResult)
+{
+    std::vector<uint256> prevIDs;
+    std::vector<uint256> finalIDs;
```

Comment:
this should _not_ be `static`, so that it can be re-used within clio! we don't want to have two implementations of the same thing.

also again, I'm not sure this is the right place for this sort of helper cc @scottschurr

Reply:
By all means, structure things so clio can reuse the code!  Given what is visible in this code review, making these functions `static` made lots of sense.  clio is outside my field of view.  Please do accommodate clio's needs.

I am a little surprised that clio can reuse any of this code, since clio uses a different library for JSON handling.  But if you can get sharing code with clio to work without bending over backwards, please do.

If you do restructure things so clio can reuse some of this code, please leave a comment where that happens.  People maintaining rippled have no idea what clio is using.  If you don't leave a comment it could be refactored away.